Ensure stacked comparison page is logging events correctly

olaseni opened the pull request:
This PR primarily ensures that tracking events for email comparison views are being sent correctly, especially for views of the `EmailProvidersStackedComparison` component. As part of this, we ensure that the event `calypso_email_providers_comparison_page_view` is sent for all comparison views, including the stacked email comparison view.

#### Changes proposed in this Pull Request

* Adds optionality to some props of `EmailProvidersStackedComparison` to ensure that earlier invocations are now compliant
* Rectifies the import alias of the stacked comparison view from `EmailProvidersComparisonStacked` to `EmailProvidersStackedComparison`
* Exchanges the `PageViewTracker` for a `TrackComponentView` for component `EmailProvidersStackedComparison`
* Adds a new `EmailProvidersStackedComparisonPage` component to host the comparison view and track page views
* Adds page view tracking for the inbox home

#### Test Preparation

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* This PR works on and establishes track events, so setup up your environment to discover/monitor track events as they are fired. There are a number of FG guides that show how to do this: PCYsg-rdg-p2
* Get a site and a single domain without an email subscription if you don't have one. We are essentially testing the email comparison page and that requires a domain without an email subscription.


#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Checkout the branch or use [calypso-live](https://github.com/Automattic/wp-calypso/pull/60631#issuecomment-1024468645) 
* Navigate to [ **Upgrades** -> **Emails** ]. 
* Confirm that you are presented with the stacked email comparison page.
* Confirm that `calypso_page_view` and `calypso_email_providers_comparison_page_view` events are triggered.
* Navigate to [ **Upgrades** -> **Domains**  ].  Click on **+Add** in the `Email` column. 
* Confirm that you are presented with the stacked email comparison page.
* Confirm that `calypso_page_view` and `calypso_email_providers_comparison_page_view` events are triggered.
* Navigate to [ **Inbox** ]. You should be presented with the stacked email comparison page.
* Confirm that you are presented with the stacked email comparison page.
* Confirm that `calypso_page_view` and `calypso_email_providers_comparison_page_view` events are triggered.


<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

olaseni committed:
Fix import name for EmailProvidersStackedComparison
* additionally add optionality to type names

olaseni committed:
Add explicit page mount for `EmailProvidersStackedComparison`

olaseni committed:
Add page view tracker for the inbox

olaseni committed:
Add page view tracker for email home

stephanethomas commented on client/my-sites/email/email-management/email-home.jsx:
```diff
@@ -82,7 +83,7 @@ class EmailManagementHome extends Component {
+					<EmailProvidersStackedComparison
 						comparisonContext="email-home-selected-domain"
 						selectedDomainName={ selectedDomainName }
 						source={ source }
```
Should we also enclose that component with a page component like you did below?

```javascript
return (
  <>
    <PageViewTracker
      path={ emailManagement( ':site', selectedDomainName ?? null ) }
      title={ 'Email Home' }
      properties={ { source, context: 'email-home-selected-domain' } }
    />
    
    <EmailProvidersStackedComparison
      comparisonContext="email-home-selected-domain"
      selectedDomainName={ selectedDomainName }
      source={ source }
    />
  </>
);
```

olaseni replied:
Yes please

olaseni replied:
Done

stephanethomas commented on client/my-sites/email/email-management/email-home.jsx:
```diff
@@ -127,6 +137,7 @@ class EmailManagementHome extends Component {
 		}
 
 		return this.renderContentWithHeader(
+			'mail-list',
```
I would rename `mail-list` to `email-list`.

stephanethomas commented on client/my-sites/email/email-management/email-home.jsx:
```diff
@@ -108,11 +111,18 @@ class EmailManagementHome extends Component {
+						title={ 'Email Home' }
+						properties={ { source, context: 'email-home-single-domain' } }
+					/>
+					<EmailProvidersStackedComparison
```
Could we add a blank line above so this block of code is easier to read? Thanks!

stephanethomas commented on client/my-sites/email/email-providers-stacked-comparison/page.tsx:
```diff
@@ -0,0 +1,26 @@
+					provider: props.selectedEmailProviderSlug,
+				} }
+			/>
+			<EmailProvidersStackedComparison { ...props } />
```
Could we add a blank line above as well 🙏?

stephanethomas commented on client/my-sites/email/email-management/email-home.jsx:
```diff
@@ -18,8 +19,8 @@ import EmailListActive from 'calypso/my-sites/email/email-management/home/email-
 import EmailPlan from 'calypso/my-sites/email/email-management/home/email-plan';
-import EmailProvidersComparisonStacked from 'calypso/my-sites/email/email-providers-stacked-comparison';
-import { emailManagementTitanSetUpMailbox } from 'calypso/my-sites/email/paths';
+import EmailProvidersStackedComparison from 'calypso/my-sites/email/email-providers-stacked-comparison';
```
Nice catch.

stephanethomas reviewed (approved):
I've left some comments but otherwise I don't have big concern (I'm worried about the duplicate pageview events that don't match reality though): 

> Navigate to [ Upgrades -> Domains -> Emails ].

The `calypso_page_view` event does not have a context in that particular case, is it expected? I mean, do we really need to add `context` to page view events or only to component view events?

![image](https://user-images.githubusercontent.com/594356/151810296-e431ccf2-26ad-4411-86ee-1ace6a3e46bd.png)

I've noticed that we trigger two `calypso_page_view` event when showing the `Inbox` page - when there is at least one mailbox:

![image](https://user-images.githubusercontent.com/594356/151810896-cd151abd-c4a2-461c-867a-e5b0e0563ff9.png)

Same thing when there is no mailbox, except that this time those events have a different `context` property:

![image](https://user-images.githubusercontent.com/594356/151811566-3375330f-3550-4b9c-b7bb-9c0ddd1fdb4b.png)

Interestingly the `path` is also different whereas the url is `/inbox`, and doesn't change.

> In the Inbox scenario, page view events are pushed twice. I see this as somewhat akin to a redirection, where each page is tracked.

I see your comment now. I agree we could see that as a redirection. However, the url in the browser doesn't change in that case. So either we log one event, or we log two but we indeed redirect the user to `/email`.

olaseni committed:
Remove duplicate events

olaseni committed:
Add page for `EmailHome`

olaseni committed:
Remove redundant argument for renderer

olaseni commented:
> > Navigate to [ Upgrades -> Domains -> Emails ].
> 
> The `calypso_page_view` event does not have a context in that particular case, is it expected? I mean, do we really need to add `context` to page view events or only to component view events?

That's intentional. The domain upsell does not need the context as there's only one view. 

I have also updated the instructions. I think there's some miscommunication. I was referring to the scenario where you navigate to  [ Upgrades -> Domains  ], the use the **+ Add** button to purchase Email.

> 
> I've noticed that we trigger two `calypso_page_view` event when showing the `Inbox` page - when there is at least one mailbox:
> 
Fixed! The selection list has its own tracker, and I missed it.
> 
> Same thing when there is no mailbox, except that this time those events have a different `context` property.
> Interestingly the `path` is also different whereas the url is `/inbox`, and doesn't change.
> 
Fixed as well. I have used an explicit page component.

> > In the Inbox scenario, page view events are pushed twice. I see this as somewhat akin to a redirection, where each page is tracked.
> 
> I see your comment now. I agree we could see that as a redirection. However, the url in the browser doesn't change in that case. So either we log one event, or we log two but we indeed redirect the user to `/email`.

This is moot because of the page component added.

olaseni committed:
Remove redundant wrappers

olaseni committed:
Remove other redundant wrappers

stephanethomas commented on client/my-sites/email/email-management/page.jsx:
```diff
@@ -0,0 +1,18 @@
+	);
+};
+
+export default EmailManagementHomePage;
```
I like this idea of suffixing page components with `Page` a lot.

stephanethomas replied:
However, could we make this new component a TypeScript component?

daledupreez replied:
Agreed on both points!

olaseni replied:
Sure

stephanethomas commented on client/my-sites/email/email-management/page.jsx:
```diff
@@ -0,0 +1,18 @@
+	return (
+		<>
+			<PageViewTracker
+				path={ emailManagement( ':site', selectedDomainName ? ':domain' : null ) }
```
Why do we pass the domain name here, we don't do that anywhere else when calling `<PageViewTracker/>`?

olaseni replied:
I'm not actually passing the domain itself, more like the domain path marker.
The path could be different based on if the selectDomainName is set.

daledupreez replied:
Good catch - this should always be `:domain`.

daledupreez replied:
Actually, you are right, 'Seni. This can be `:domain` or null.

stephanethomas replied:
First, why would `selectedDomainName` be null? If we don't specify a domain name in the url (i.e. https://wordpress.com/email) then we are presented with the site picker, not this page. Second, we don't perform that check anywhere else: why do we want to have a specific behavior for that particular page?

daledupreez replied:
There is another case where we show this page: `/email/:siteSlug`
For that case, the domain isn't specified, but we do have the site.

stephanethomas replied:
Ah right, my bad. I'm not sure what I was thinking this morning 🤦‍♂️.

daledupreez commented on client/my-sites/email/email-management/page.jsx:
```diff
@@ -0,0 +1,18 @@
+import EmailManagementHome from 'calypso/my-sites/email/email-management/email-home';
+import { emailManagement } from 'calypso/my-sites/email/paths';
+
+const EmailManagementHomePage = ( { source, selectedDomainName } ) => {
```
Nit: it may be more appropriate to call this file `home-page` instead of just `page`, as we are currently losing the "Home" part of the component name.

daledupreez commented on client/my-sites/email/email-providers-stacked-comparison/page.tsx:
```diff
@@ -0,0 +1,26 @@
+import { ReactElement } from 'react';
```
Nit:`ReactElement` should be imported as a type.

You should also be able to use `JSX.Element` without an import, but that's even more of a nit-pick.

daledupreez commented on client/my-sites/email/email-providers-stacked-comparison/page.tsx:
```diff
@@ -0,0 +1,26 @@
+import PageViewTracker from 'calypso/lib/analytics/page-view-tracker';
+import EmailProvidersStackedComparison from 'calypso/my-sites/email/email-providers-stacked-comparison/index';
+import { emailManagementPurchaseNewEmailAccount } from 'calypso/my-sites/email/paths';
+import type { EmailProvidersStackedComparisonProps } from 'calypso/my-sites/email/email-providers-stacked-comparison/index';
```
Question: is it necessary to include the trailing `/index` in the import path?

olaseni replied:
The IDE pulled that in. I'm fixing it.

daledupreez reviewed (commented):
Dropping initial comments for now -- will be adding testing notes shortly.

daledupreez reviewed (approved):
Overall, this is testing well and LGTM! The only possible item is to switch the one component to TS.

Beyond that, the only oddity I found is that our logic for showing the email upsell from the Inbox is _really_ roundabout: we load the `EmailManagementHome` component, which in turn makes the decision to show an upsell or the domain list depending on state. That seems like it could be a good candidate for a refactor in future!

#### Evidence of testing

<img width="845" alt="Screenshot 2022-02-02 at 12 30 06" src="https://user-images.githubusercontent.com/3376401/152136983-0221510d-72bf-4435-87b0-d6e406e97abd.png">
<img width="845" alt="Screenshot 2022-02-02 at 12 29 50" src="https://user-images.githubusercontent.com/3376401/152136996-3ee919a2-d6f8-4243-b0e3-35624a503a37.png">
<img width="845" alt="Screenshot 2022-02-02 at 12 29 06" src="https://user-images.githubusercontent.com/3376401/152137004-4e581a12-f753-44d0-9d7a-188e0706175e.png">

olaseni committed:
Convert page to TSX

olaseni committed:
Trim fat from TSX

daledupreez reviewed (approved):
I've just re-tested and everything LGTM!

I'll get this merged as 'Seni is AFK today.
